draft-ietf-masque-h3-datagram.md

```diff
@@ -280,11 +291,75 @@ yet received a HEADERS frame, the endpoint MUST abruptly terminate the
 corresponding stream with a stream error of type H3_GENERAL_PROTOCOL_ERROR.
 
 Servers MUST NOT send a REGISTER_DATAGRAM_CONTEXT capsule on a stream before
-they have received at least one REGISTER_DATAGRAM_CONTEXT capsule from the
-client on that stream. This ensures that clients control whether datagrams are
-allowed for a given request. If a client receives a REGISTER_DATAGRAM_CONTEXT
-capsule on a stream where the client has not yet sent a
-REGISTER_DATAGRAM_CONTEXT capsule, the client MUST abruptly terminate the
+they have received at least one REGISTER_DATAGRAM_CONTEXT capsule or one
+REGISTER_DATAGRAM_NO_CONTEXT capsule from the client on that stream. This
+ensures that clients control whether datagrams are allowed for a given request.
+If a client receives a REGISTER_DATAGRAM_CONTEXT capsule on a stream where the
+client has not yet sent a REGISTER_DATAGRAM_CONTEXT capsule, the client MUST
+abruptly terminate the corresponding stream with a stream error of type
+H3_GENERAL_PROTOCOL_ERROR.
+
+Servers MUST NOT send a REGISTER_DATAGRAM_CONTEXT capsule on a stream where it
+has received a REGISTER_DATAGRAM_NO_CONTEXT capsule. If a client receives a
+REGISTER_DATAGRAM_CONTEXT capsule on a stream where the client has sent a
+REGISTER_DATAGRAM_NO_CONTEXT capsule, the client MUST abruptly terminate the
+corresponding stream with a stream error of type H3_GENERAL_PROTOCOL_ERROR.
+
+
+## The REGISTER_DATAGRAM_NO_CONTEXT Capsule {#register-no-context-capsule}
+
+The REGISTER_DATAGRAM_NO_CONTEXT capsule (type=0x03) allows a client to inform
+the server that datagram contexts will not be used with this stream. It also
+informs the server of the encoding and semantics of datagrams associated with
+this stream. Its Capsule Data field consists of:
+
+~~~
+REGISTER_DATAGRAM_NO_CONTEXT Capsule {
+  Extension String (..),
+}
+~~~
+{: #register-no-context-capsule-format title="REGISTER_DATAGRAM_NO_CONTEXT Capsule Format"}
+
+Extension String:
+
+: A string of comma-separated key-value pairs to enable extensibility, see
+the definition of the same field in {{register-capsule}} for details.
+
+Note that this registration is unilateral and bidirectional: the client
+unilaterally defines the semantics it will apply to the datagrams it sends and
+receives with this stream.
+
+Endpoints MUST NOT send DATAGRAM frames without a Context ID until they have
+either sent or received a REGISTER_DATAGRAM_NO_CONTEXT Capsule. However, due to
+reordering, an endpoint that receives a DATAGRAM frame before receiving either
+a REGISTER_DATAGRAM_CONTEXT capsule or a REGISTER_DATAGRAM_NO_CONTEXT capsule
+MUST NOT treat it as an error, it SHALL instead drop the DATAGRAM frame
+silently, or buffer it temporarily while awaiting a
+REGISTER_DATAGRAM_NO_CONTEXT capsule or the corresponding
+REGISTER_DATAGRAM_CONTEXT capsule.
+
+Servers MUST NOT send the REGISTER_DATAGRAM_NO_CONTEXT capsule. If a client
```

Comment:
As far as I know, HTTP/3 doesn't have any other way to abruptly close request streams apart from reseting the stream. That reset will be propagated, and if the error code is changed that's not a big deal. This matches how HTTP/3 CONNECT reacts to a TCP RST for example.

Reply:
It's good behaviour to propagate errors across hops, even if they require mapping. For example, see the H2-H3 mapping advice in https://datatracker.ietf.org/doc/html/draft-ietf-quic-http-34#appendix-A.4.1.

I think the text is fine right to focus on HTTP/3 right now. Subsequent follow up work to make this design work over other HTTP versions can add guidance on how to propagate errors if specific advice is needed. I wouldn't worry about it isolation since we might overlook details.